pep-0612.rst

```diff
@@ -349,26 +554,82 @@ three categories (positional-only, positional-or-keyword, keyword-only) we’re
 mentioned when discussing ``.args`` and ``.kwargs``. Fundamentally, in order to
 capture two categories when there are some things that can be in either
-category, we need a higher level primitive (\ ``ParameterSpecification``\ ) to
+category, we need a higher level primitive (\ ``ParamSpec``\ ) to
 capture all three, and then split them out afterward.
 
-Mutations on ParameterSpecifications
-^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+Defining ParametersOf
+^^^^^^^^^^^^^^^^^^^^^^
 
-There are still a class of decorators still not supported with these features:
-those that mutate (add/remove/change) the parameters of the given function.
-Defining operators that do these mutations becomes very complicated very
-quickly, as you have to deal with name collision issues much more prominently.
-We will defer that work until there is significant demand, and then we would be
-open to revisiting it.
+Another proposal we considered was defining ``ParametersOf`` and ``ReturnType``
+operators which would operate on a domain of a newly defined ``Function`` type.
+``Function`` would be callable with, and only with ``ParametersOf[F]``.
+``ParametersOf`` and ``ReturnType`` would only operate on type variables with
+precisely this bound.  The combination of these three features could express
+everything that we can express with ``ParamSpecs``.
+
+
+.. code-block::
+
+   F = TypeVar("F", bound=Function)
+
+   def no_change(f: F) -> F:
+       def inner(
+         *args: ParametersOf[F].args,
+         **kwargs: ParametersOf[F].kwargs
+       ) -> ReturnType[F]:
+          return f(*args, **kwargs)
+       return inner
+
+   def wrapping(f: F) -> Callable[ParametersOf[F], List[ReturnType[F]]]:
+       def inner(
+           *args: ParametersOf[F].args,
+           **kwargs: ParametersOf[F].kwargs
+        ) -> List[ReturnType[F]]:
+          return [f(*args, **kwargs)]
+       return inner
+
+   def unwrapping(
+       f: Callable[ParametersOf[F], List[TReturn]]
+   ) -> Callable[ParametersOf[F], TReturn]:
+       def inner(
+          *args: ParametersOf[F].args,
+          **kwargs: ParametersOf[F].kwargs
+       ) -> TReturn:
+          return f(*args, **kwargs)[0]
+       return inner
+
+We decided to go with ``ParamSpec``\ s over this approach for several reasons:
+
+* The footprint of this change would be larger, as we would need two new
+  operators, and a new type, while ``ParamSpec`` just introduces a new variable.
+* Python typing has so far has avoided supporting operators, whether
+  user-defined or built-in, in favor of destructuring.  Accordingly,
+  ``ParamSpec`` based signatures look much more like existing Python.
+* The lack of user-defined operators makes common patterns hard to spell.
+  ``unwrapping`` is odd to read because ``F`` is not actually referring to any
+  callable. It’s just being used as a container for the parameters we wish to
+  propagate.  It would read better if you could define an operator
```

Comment:
nit: replace the second-person pronouns ("you", "your") with first-person pronouns ("we", "our") for consistency with the context?